dataset/batch_image.py

```diff
@@ -1,525 +1,767 @@
+                def __wrapper():
+                    #pylint: disable=cell-var-from-loop
+                    wrapped_method = method
+                    def __func(self, *args, src='images', dst='images', **kwargs):
+                        return getattr(cls, wrapper)(self, wrapped_method, src=src, dst=dst,
+                                                     use_self=True, *args, **kwargs)
+                    return __func
+                name_slice = slice(len(prefix), -len(suffix))
+                wrapped_method_name = method_name[name_slice]
+                setattr(cls, wrapped_method_name, action(__wrapper()))
+        return cls
+    return __decorator
 
-    def get_image_size(self, image):
-        """ Return an image size (width, height) """
-        raise NotImplementedError("Should be implemented in child classes")
 
-    @action
-    def crop(self, components='images', origin=None, shape=None):
-        """ Crop all images in the batch
+class BaseImagesBatch(Batch):
+    """ Batch class for 2D images """
+    components = "images", "labels"
+    #pylint: disable=arguments-differ
+    def _assemble(self, all_res, *args, components='images', **kwargs):
+        """ Assemble the batch after a parallel action.
 
         Parameters
         ----------
-        components : str
-            a component name or names which data should be cropped
+        all_res : list
+            processed components of the batch
 
-        origin : tuple
-            can be one of:
+        components : str, sequence
+            names of the components to assemble
 
-            - tuple - a starting point in the form of (x, y)
-            - CROP_00 - to crop from left top edge (0,0)
-            - CROP_CENTER - to crop from center of each image
+        Returns
+        -------
+        self
+        """
+        _ = all_res, args, kwargs, components
+        raise NotImplementedError("Must be implemented in a child class")
 
-        shape : tuple
-            a crop size in the form of (width, height)
+    def _make_path(self, path, ix):
+        """ Compose path.
+
+        Parameters
+        ----------
+        path : str, None
+        ix : str
+            element's index (filename)
+
+        Returns
+        -------
+        path : str
+            joined path if path is not None else element's path specified in the batch's index
         """
-        if origin is not None or shape is not None:
-            origin = origin if origin is not None else (0, 0)
-            self._crop(components, origin, shape)
-        return self
+        return self.index.get_fullpath(ix) if path is None else os.path.join(path, ix)
 
-    @action
-    def random_crop(self, components='images', shape=None):
-        """ Crop all images to a given shape and a random origin
+    def _load_image(self, ix, src=None, components="images"):
+        """ Loads image
+
+        .. note:: only works with a single component
```

Comment:
Single component in the batch class? Or where? What exactly works?

In fact, you don't need tis note as parameters section clearly describes that `components` can be `str` only.

To bring attention to this fact you might write a note like "Take into account that `components` can be `str` only, `array-like` is not allowed here".

Reply:
+